File: tests/exr-interoperability.py
"""Independent compressed, multipart and mixed-channel OpenEXR verification."""
import json,sys
from pathlib import Path
import numpy as np
import OpenEXR
out=Path(sys.argv[1] if len(sys.argv)>1 else 'test-results/gpu/interop')
report={'implementation':'OpenEXR official Python module','version':getattr(OpenEXR,'__version__','unknown'),'checks':[]}
for fixture in json.loads((out/'aureon-generated.json').read_text()):
    with OpenEXR.File(str(out/fixture['file']),separate_channels=True) as image:
        assert len(image.parts)==len(fixture['parts'])
        for i,part in enumerate(fixture['parts']):
            for name,expected in part['channels'].items():
                actual=image.parts[i].channels[name].pixels
                assert actual.shape==(part['height'],part['width'])
                np.testing.assert_allclose(actual.reshape(-1),expected,rtol=0,atol=.002)
                if name=='objectId':assert actual.dtype==np.uint32
    report['checks'].append({'name':'Official OpenEXR decodes '+fixture['file'],'passed':True})
# Reciprocal test: independent encoder -> Aureon decoder. Multiple scanline
# blocks (height 19), compressible values, fractional HDR, UINT and HALF mixed.
manifest=[]
for compression_name in ['ZIP','ZIPS']:
    channels={'R':(np.arange(37*19).reshape(19,37)%37/8-1).astype('float16'),'G':np.full((19,37),4.5,dtype='float32'),'objectId':np.arange(37*19,dtype='uint32').reshape(19,37)+4000000000}
    for multi in [False,True]:
        header={'compression':getattr(OpenEXR,compression_name+'_COMPRESSION'),'type':OpenEXR.scanlineimage}
        parts=[OpenEXR.Part(header,{'R':channels['R'],'G':channels['G']},'beauty'),OpenEXR.Part(header,{'objectId':channels['objectId']},'guides')] if multi else [OpenEXR.Part(header,channels,'part0')]
        name=f'openexr-{compression_name}-{multi}.exr'
        with OpenEXR.File(parts) as image:image.write(str(out/name))
        manifest.append({'file':name,'parts':[{'width':37,'height':19,'channels':{k:v.pixels.reshape(-1).tolist() for k,v in p.channels.items()}} for p in parts]})
(out/'openexr-generated.json').write_text(json.dumps(manifest))
(out/'report.json').write_text(json.dumps(report,indent=2)+'\n')
print(json.dumps(report,indent=2))

File: tests/exr-interoperability.mjs
/** Cross-implementation fixtures: Node writes Aureon EXR; Python uses OpenEXR.
 * Expected values are saved separately; comparisons do not use Aureon's decoder.
 */
import {mkdir,writeFile,readFile} from 'node:fs/promises';
import {join} from 'node:path';
import assert from 'node:assert/strict';
import {encodeEXRAdvanced,decodeEXRAdvanced} from '../src/io/exr-advanced.js';
const out=process.argv[3]||'test-results/gpu/interop';await mkdir(out,{recursive:true});
if(process.argv[2]==='verify'){
 const manifest=JSON.parse(await readFile(join(out,'openexr-generated.json'),'utf8'));
 for(const entry of manifest){const decoded=await decodeEXRAdvanced(new Uint8Array(await readFile(join(out,entry.file))));assert.equal(decoded.parts.length,entry.parts.length);for(let i=0;i<entry.parts.length;i++){const expected=entry.parts[i],actual=decoded.parts[i];assert.equal(actual.width,expected.width);assert.equal(actual.height,expected.height);for(const [name,values]of Object.entries(expected.channels)){assert.equal(actual.channels[name].length,values.length);actual.channels[name].forEach((v,j)=>assert.ok(Math.abs(v-values[j])<.003,`${entry.file} part ${i} ${name} pixel ${j}: ${v} != ${values[j]}`));}}}
 console.log('PASS',manifest.length,'independently encoded EXRs decoded by Aureon');
}else{
 const width=37,height=19,n=width*height,channels={R:Float32Array.from({length:n},(_,i)=>(i%37-12)/8),G:new Float32Array(n).fill(4.5),B:new Float32Array(n).fill(-.25),objectId:Uint32Array.from({length:n},(_,i)=>4000000000+i),depth:Float32Array.from({length:n},(_,i)=>i/4)};
 const fixtures=[];
 for(const compression of ['NONE','ZIP','ZIPS'])for(const multipart of [false,true]){
  const file=`aureon-${compression}-${multipart?'multi':'single'}.exr`,base={width,height,types:{R:'HALF',G:'FLOAT',B:'HALF',objectId:'UINT',depth:'FLOAT'},compression},parts=multipart?[{...base,name:'beauty',channels:{R:channels.R,G:channels.G,B:channels.B}},{...base,name:'guides',channels:{objectId:channels.objectId,depth:channels.depth}}]:[{...base,name:'part0',channels}];
  await writeFile(join(out,file),await encodeEXRAdvanced(multipart?{parts}:parts[0]));fixtures.push({file,parts:parts.map(p=>({...p,channels:Object.fromEntries(Object.entries(p.channels).map(([k,v])=>[k,Array.from(v)]))}))});
 }
 await writeFile(join(out,'aureon-generated.json'),JSON.stringify(fixtures));console.log('Wrote',fixtures.length,'cross-decoder fixtures');
}
